### PR TITLE
[alertmanager] add new chart

### PR DESCRIPTION
**Pull request by naseemkullah**
#### Special notes for your reviewer:

Adds a minimalist alertmanager chart.

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Comment by torstenwalter** [comment]
Did you see that AlertManager is part of https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus-operator/templates/alertmanager?

**Comment by scottrigby** [comment]
It would be good to provide some context. I happen to know from earlier conversations with @naseemkullah and @brancz that this is a step toward componentizing that chart into standalone parts. In future, the kube-prometheus-stack chart could switch to pulling this in as a dependency. Alternatively this could be installed separately with other standalone charts like prometheus.

But @naseemkullah you might have more context details.

**Comment by naseemkullah** [comment]
Use cases would indeed be for standalone chart and potentially eventual replacement of alertmanager component in prometheus chart (not prometheus-operator/kube-prometheus-stack chart though, since it is a CRD there).

**Review by Xtigyro — CHANGES_REQUESTED** [review]
Why would anybody need a separate chart for Alertmanager?

**Comment by naseemkullah** [comment]
@Xtigyro 

One may choose to use alertmanager as the alerting system for a client other than prometheus.

One may choose to deploy multiple instances of alertmanager to obtain HA.

For systems that deploy a vendored Prometheus but not Alertmanger (namely Linkerd see: https://github.com/linkerd/linkerd2/pull/4220/files) in which case it would be nice to have a community alertmanager chart to deploy alongside it.

**Comment by Xtigyro** [comment]
I'm torn here.

@zanhsieh @monotek @bismarck @vsliouniaev @gianrubio 
Your thoughts please.

**Review by monotek — CHANGES_REQUESTED** [review]

**Comment by monotek** [comment]
@Xtigyro 

Generally i like the idea to have alertmanager as standalone chart, as it can be used as dependency in other charts or you could solve one of the usecases mentioned by @naseemkullah

**Comment by Xtigyro** [comment]
@naseemkullah Would you be able to work on splitting the Alertmanager component from the `prometheus` chart and using this as a dependency?

**Comment by naseemkullah** [comment]
> @naseemkullah Would you be able to work on splitting the Alertmanager component from the `prometheus` chart and using this as a dependency?

Eventually yes, not immediately though, I also think we should discuss the best way forward in terms of feature parity. There is a lot of cruft we should reevaluate in the prometheus chart('s subcomponents). It would not be a bad idea to have a prometheus maintainers call at some point to figure out what we want to do.

**Comment by Xtigyro** [comment]
@naseemkullah Very nice! After resolving the comments from @monotek - I believe this should be good to go.

**Comment by monotek** [comment]
Maybe it would also make sense to add the same maintainers for the standalone charts as for the prometheus chart?

At least when the standalone chart(s) hould used as dependencies in the prometheus chart later?

Having a call about this sounds good to me too. Not sure if dependency charts should replace current implementation or be usable optionally. So let's have a call :-)

What's your timezone guys? I'm in cet.

**Comment by naseemkullah** [comment]
> Maybe it would also make sense to add the same maintainers for the standalone charts as for the prometheus chart?
> 
> At least when the standalone chart(s) should be used as dependencies in the prometheus chart later?
> 
> Having a call about this sounds good to me too. Not sure if dependency charts should replace current implementation or be usable optionally. So let's have a call :-)
> 
> What's your timezone guys? I'm in cet.

EST, we can start with an async group chart in k8s slack too

**Comment by scottrigby** [comment]
FWIW I'm really encouraged by this thread

**Comment by GMartinez-Sisti** [comment]
I'm sure you all know this already, but I would like to remind that the Alertmanager deployed in kube-prometheus-stack is managed by the prometheus-operator operator, so I believe this chart does not qualify as a possible dependency for that chart, but instead, as already mentioned, an isolated solution for other use cases.

**Comment by naseemkullah** [comment]
> I'm sure you all know this already, but I would like to remind that the Alertmanager deployed in kube-prometheus-stack is managed by the prometheus-operator operator, so I believe this chart does not qualify as a possible dependency for that chart, but instead, as already mentioned, an isolated solution for other use cases.

Yep this may have gotten buried in the other comments: https://github.com/prometheus-community/helm-charts/pull/64#issuecomment-688677072

**Comment by naseemkullah** [comment]
Thanks for the great feedback @monotek , changes applied and ready for re-review.

@zanhsieh @gianrubio @Xtigyro  please let me know if you'd like to help maintain this chart as well and I will add you to the maintainers list. Thanks.

File: charts/alertmanager/values.yaml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  ## Persistent Volume Storage Class
+  ## If defined, storageClassName: <storageClass>
+  ## If set to "-", storageClassName: "", which disables dynamic provisioning
+  ## If undefined (the default) or set to null, no storageClassName spec is
+  ## set, choosing the default provisioner.
+  ##
+  # storageClass: "-"
+  accessModes:
+    - ReadWriteOnce
+  size: 8Gi
```
[the remark this answers is not in the document]
Comment (naseemkullah) [reply charts/alertmanager/values.yaml]: 1GB it is

File: charts/alertmanager/values.yaml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  #      - alertmanager.domain.com
+
+resources: {}
+  # We usually recommend not to specify default resources and to leave this as a conscious
+  # choice for the user. This also increases chances charts run on environments with little
+  # resources, such as Minikube. If you do want to specify resources, uncomment the following
+  # lines, adjust them as necessary, and remove the curly braces after 'resources:'.
+  # limits:
+  #   cpu: 100m
+  #   memory: 128Mi
```
[the remark this answers is not in the document]
Comment (naseemkullah) [reply charts/alertmanager/values.yaml:60]: 32Mi it is

File: charts/alertmanager/values.yaml [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  # Annotations to add to the service account
+  annotations: {}
+  # The name of the service account to use.
+  # If not set and create is true, a name is generated using the fullname template
+  name:
+
+podSecurityContext: {}
+  # fsGroup: 2000
+
+securityContext: {}
```
[the remark this answers is not in the document]
Comment (naseemkullah) [reply charts/alertmanager/values.yaml]: Thanks, used defaults from vendored alertmanager component in prom chart, though fsgroup was moved to pod level

File: charts/alertmanager/values.yaml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  # Specifies whether a service account should be created
+  create: true
+  # Annotations to add to the service account
+  annotations: {}
+  # The name of the service account to use.
+  # If not set and create is true, a name is generated using the fullname template
+  name:
+
+podSecurityContext: {}
+  # fsGroup: 2000
```
[the remark this answers is not in the document]
Comment (naseemkullah) [reply charts/alertmanager/values.yaml]: See https://github.com/prometheus-community/helm-charts/pull/64#discussion_r487551665

File: charts/alertmanager/values.yaml [hunk 5]
```diff
@@ -0,0 +1,97 @@
+# Default values for alertmanager.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+replicaCount: 1
+
+image:
+  repository: prom/alertmanager
+  pullPolicy: IfNotPresent
```
[the remark this answers is not in the document]
Comment (naseemkullah) [reply charts/alertmanager/values.yaml:9]: Now added as per https://github.com/helm/helm/blob/e2442699fa4703456b16884990c5218c16ed16fc/pkg/chartutil/create.go#L103

File: charts/alertmanager/values.yaml [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  # runAsUser: 1000
+
+service:
+  type: ClusterIP
+  port: 9093
+
+ingress:
+  enabled: false
+  annotations: {}
+    # kubernetes.io/ingress.class: nginx
```
[the remark this answers is not in the document]
Comment (naseemkullah) [reply charts/alertmanager/values.yaml]: TIL, thanks, removed. we should make a PR to remove from https://github.com/helm/helm/blob/e2442699fa4703456b16884990c5218c16ed16fc/pkg/chartutil/create.go#L138

File: charts/alertmanager/Chart.yaml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v2
+name: alertmanager
+description: The Alertmanager handles alerts sent by client applications such as the Prometheus server.
+home: https://prometheus.io/
+icon: https://raw.githubusercontent.com/prometheus/prometheus.github.io/master/assets/prometheus_logo-cb55bb5c346.png
+sources:
+  - https://github.com/prometheus/alertmanager
+type: application
+version: 0.0.1
```
[the remark this answers is not in the document]
Comment (naseemkullah) [reply charts/alertmanager/Chart.yaml]: TIL thanks

File: charts/alertmanager/templates/NOTES.txt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  export NODE_IP=$(kubectl get nodes --namespace {{ .Release.Namespace }} -o jsonpath="{.items[0].status.addresses[0].address}")
+  echo http://$NODE_IP:$NODE_PORT
+{{- else if contains "LoadBalancer" .Values.service.type }}
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get --namespace {{ .Release.Namespace }} svc -w {{ include "alertmanager.fullname" . }}'
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ include "alertmanager.fullname" . }} --template "{{"{{ range (index .status.loadBalancer.ingress 0) }}{{.}}{{ end }}"}}")
+  echo http://$SERVICE_IP:{{ .Values.service.port }}
+{{- else if contains "ClusterIP" .Values.service.type }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app.kubernetes.io/name={{ include "alertmanager.name" . }},app.kubernetes.io/instance={{ .Release.Name }}" -o jsonpath="{.items[0].metadata.name}")
+  echo "Visit http://127.0.0.1:8080 to use your application"
```
[the remark this answers is not in the document]
Comment (naseemkullah) [reply charts/alertmanager/templates/NOTES.txt]: added template, should do the same via PR for https://github.com/helm/helm/blob/e2442699fa4703456b16884990c5218c16ed16fc/pkg/chartutil/create.go#L385-L386

**Review by monotek — DISMISSED** [review]

**Comment by Xtigyro** [comment]
@scottrigby Shouldn't we create a PR for the Prometheus chart before merging this one - so no double efforts are required for maintenance of Alertmanager?

**Comment by monotek** [comment]
@Xtigyro 
I don't think so.
Imho we don't plan to change anything in the altermanager component for prometheus so i don't seee double effort there.
It could also be a benefit for others, if they can use it now as they won't need to do work twice.
Standalone usage is also an argument to merge it now.

File: charts/alertmanager/README.md [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# alertmanager
+
+Chart to deploy alertmanager as a statefulset.
+
+## Installation
+
+```sh
```
Comment (scottrigby) [inline charts/alertmanager/README.md:7]: My only thought now is to improve the README a bit. How about the README template laid out for the other charts in this repo?

**Review by scottrigby — CHANGES_REQUESTED** [review]
Just this one suggestion. Otherwise this looks good!

**Comment by scottrigby** [comment]
Re dependency, we are thinking this could be a drop in replacement for the alert manager in the prometheus chart, right?

If so, I was thinking this is the order:

1. This new alertmanager chart PR must be merged in order to create the chart version package
2. then a PR can be made against the prometheus chart to replace the existing component with this newly available alertmanager package as a dependency

But either way, that would be a follow up PR. Am I right this is what you're thinking though @naseemkullah ?

**Comment by Xtigyro** [comment]
> 
> 
> Re dependency, we are thinking this could be a drop in replacement for the alert manager in the prometheus chart, right?
> 
> If so, I was thinking this is the order:
> 
>     1. This new alertmanager chart PR must be merged in order to create the chart version package
> 
>     2. then a PR can be made against the prometheus chart to replace the existing component with this newly available alertmanager package as a dependency
> 
> 
> But either way, that would be a follow up PR. Am I right this is what you're thinking though @naseemkullah ?

@scottrigby @naseemkullah 

OK - very nice! Agreed.
P.S. I'll be happy to help with the follow-up PR.

**Review by Xtigyro — APPROVED** [review]

File: charts/alertmanager/README.md (hunk 9 above)
Comment (naseemkullah) [reply charts/alertmanager/README.md]: Sounds good! I did not find a template per se but mimicked what I saw in prometheus adapter. PTAL

**Review by monotek — APPROVED** [review]

**Review by Xtigyro — APPROVED** [review]

**Review by scottrigby — APPROVED** [review]
🎉

**Comment by torstenwalter** [comment]
@naseemkullah @monotek Chart looks great. The only thing which is unclear for me is who maintains that chart?

In CODEOWNERS is says:
`/charts/alertmanager/ @monotek @naseemkullah`

In Chart.yaml it's just

```
maintainers:
  - name: naseemkullah
    email: naseem@transit.app
```

**Pull request merged**